### ISSUE TITLE
SWR Iterations on COSMOS not working

### ISSUE THREAD

**Issue by valentinaschueller** [valentinaschueller/ece-scm-coupling#11]
I ran into a strange issue getting the iterations on COSMOS to work. I fixed it now but I'll use this issue to document what was going on.

**Comment by valentinaschueller** [comment]
This was the issue I filed for the support:

I have encountered a problem using Python as a wrapper for a code that gets executed with mpirun multiple times in a row. In my workflow, Python sets up the runtime environment, then I launch the code from Python (which includes calling `mpirun`). After that, I postprocess some NetCDF output files using xarray and re-run the code. I had the issue that running the code after postprocessing with xarray did not work. The whole setup is very involved but I managed to come up with a simple, reproducible example to showcase the issue. It seems that the main problem is an incompatibility of the MPI versions used by mpi4py and mpirun, but I could be wrong.
 
  I provide the output of `module list` below this message.
 
  Here is what you have to do to reproduce the problem:
```bash
  > ml foss/2023a
  > ml xarray/2023.9.0
  > ml netcdf4-python/1.6.4
  > python break_mpirun_from_python.py
``` 
 
This Python script runs `mpirun -n 1 hello_world` twice (the most basic MPI Hello World program in C)[^1]. In between the calls to hello_world, it either
  a) imports and uses xarray to read in a NetCDF file (case 0);
  b) imports MPI from mpi4py (case 1);
  c) does not do anything (default).
  You can switch between these cases by setting what_to_try in l. 6 of this script.
 
  I expect this code to print "Hello, World!" twice in all cases.
  As you will see, the second call to `mpirun` does not work in cases (a) and (b). It does not print any error messages (I checked by appending `2> stderr.txt` to the command in l. 9).
  In case (c), the code prints "Hello, World!" twice, as expected. Removing `mpirun -n 1` from the command defined in l. 9 restores the expected behavior.
 

[^1]: https://mpitutorial.com/tutorials/mpi-hello-world/

**Comment by valentinaschueller** [comment]
After some discussion with our support, I managed to fix the test script:

I tried something else considering what you wrote: If I understood you correctly, I am causing a problem when calling mpirun the second time, since both mpi4py and mpirun can initialize MPI. To fix this, I have added MPI.Finalize() after "from mpi4py import MPI".

Here are the good news: I was able to run this successfully on COSMOS once! Here is the output: (Note the warning regarding mlx5_0, you might have a better idea of what this means?)

```
Case: 1
first command: mpirun -n 1 /home/vschuller/hello_world
cosmos3:rank0.hello_world: Failed to modify UD QP to INIT on mlx5_0: Operation not permitted
cosmos3:rank0.hello_world: Failed to modify UD QP to INIT on mlx5_0: Operation not permitted
cosmos3:rank0.hello_world: Failed to modify UD QP to INIT on mlx5_0: Operation not permitted
cosmos3:rank0.hello_world: Failed to modify UD QP to INIT on mlx5_0: Operation not permitted
Hello world from processor cosmos3, rank 0 out of 1 processors
first command done
made it through case 1
2nd command:  mpirun -n 1 /home/vschuller/hello_world
Hello world from processor cosmos3, rank 0 out of 1 processors
2nd command end
```

Here are the bad news: This does not work every time! Most times, the code gets stuck when calling MPI.Finalize(), the last lines being:
```
cosmos3:rank0.hello_world: Failed to modify UD QP to INIT on mlx5_0: Operation not permitted
```

I ran into these output messages before but did not really investigate what's going on (since they do not show up every time). A bit of googling brought me here:
https://github.com/easybuilders/easybuild-easyconfigs/issues/20233#issuecomment-2023410182

Since I need to make sure that xarray, netCDF4, and mpi4py all use the flags suggested, I picked the second suggestion (setting the environment variables):
```bash
export OMPI_MCA_btl='^uct,ofi'
export OMPI_MCA_pml='ucx'
export OMPI_MCA_mtl='^ofi'
```

before running the Python script. To get the xarray and netCDF4 code to run, I additionally need to add
```
from mpi4py import MPI
MPI.Finalize()
```

before continuing with calling mpirun.

**Comment by valentinaschueller** [comment]
This is the script I used to understand what's going on:

```python
import os

# case 0: try importing + using xarray between running scripts
# case 1: try importing mpi4py between running scripts
# case 2: try importing netCDF4
# else: just run the script twice
what_to_try = 1
print(f"Case: {what_to_try}")

command = "mpirun -n 1 /home/x_valsc/hello_world"
print("first command:", command)
os.system(command)
print("first command done")

if what_to_try == 0:
    # just importing xarray is no problem
    import xarray as xr
    # but as soon as we load a NetCDF file, mpirun breaks
    da = xr.open_dataarray("test.nc")
    # it does not help to close the DataArray before continuing:
    da.close()
    print(f"survived case 0")
elif what_to_try == 1:
    # just importing mpi4py is sufficient to break mpirun
    from mpi4py import MPI
    print(MPI.Is_initialized())
    MPI.Finalize()
    print(MPI.Is_finalized())
    print(f"made it through case 1")
elif what_to_try == 2:
    # just importing netCDF4 is sufficient to break mpirun
    import netCDF4
else:
    print(f"nothing more to do")
    pass

print("2nd command: ",command)
os.system(command) #afaik, this is the only line that gets ignored
print("2nd command end")
```

**Comment by valentinaschueller** [comment]
The fix in eec94c7 addresses this by finalizing MPI if it is not already finalized. Additionally, I set the environment variables in my COSMOS `.bash_profile` (this is obviously not part of this package.

**Comment by valentinaschueller** [comment]
closed by #12

**Comment by valentinaschueller** [comment]
Actually, the fix is not necessary, will revert the change. Reason:

The issue is that the netCDF4-Python package on LUNARC is built with Parallel I/O (https://unidata.github.io/netcdf4-python/#parallel-io), i.e., it comes with mpi4py and will initialize MPI whenever it is imported. xarray uses netCDF4 for loading NetCDF files, thus causing the issue we discussed before.

Since I do not need parallel I/O, the easiest way to fix it is like this:

1. Load foss, NetCDF, Python, xarray modules as usual. 
2. I created my own Python virtual environment (venv). So far, xarray is available while neither netCDF4 nor mpi4py are installed (check with pip list).
3. In the venv, I ran `pip install netCDF4`. Since mpi4py is not installed and not required, this installs netCDF4 with parallel I/O *disabled*.
4. Now both test cases (case 0, case 2) work without a problem. Case 1 fails since mpi4py is not installed, but that's expected since I don't want that package to be available.
